fuse/file.go

```diff
@@ -3,103 +3,370 @@ package fuse
+}
+
+// If currently writing a file-like object, we should use local state to fulfil many requests.
+func (f *file) useLocalState() bool {
+	return len(f.writers) != 0 && f.isFileLike()
+}
+
+var _ = fs.Node(&file{})
+var _ = fs.Handle(&file{})
+
+func (f *file) Attr(ctx context.Context, a *fuse.Attr) error {
+	f.mux.Lock()
+	defer f.mux.Unlock()
+
+	if !f.useLocalState() {
+		// Fetch updated attributes only if we're not currently writing to it.
+		entry, err := f.refind(ctx)
+		if err != nil {
+			activity.Warnf(ctx, "FUSE: Attr errored %v, %v", f, err)
+			return err
+		}
+		f.entry = entry
+	}
+
+	f.fillAttr(a)
+	activity.Record(ctx, "FUSE: Attr %v: %+v", f, *a)
+	return nil
 }
 
-// Open a file for reading or writing.
+func (f *file) fillAttr(a *fuse.Attr) {
+	attr := plugin.Attributes(f.entry)
+	applyAttr(a, attr, getFileMode(f.entry))
+
+	if f.useLocalState() || !attr.HasSize() {
+		// Use whatever size we know locally. Retrieving content can be expensive so we settle for
+		// including size only when it's been retreived previously by opening the file.
+		a.Size = f.readSize
+	}
+}
+
+func getFileMode(entry plugin.Entry) os.FileMode {
+	var mode os.FileMode
+	if plugin.WriteAction().IsSupportedOn(entry) {
+		mode |= 0220
+	}
+	if plugin.ReadAction().IsSupportedOn(entry) ||
+		plugin.StreamAction().IsSupportedOn(entry) {
+		mode |= 0440
+	}
+	return mode
+}
+
+var _ = fs.NodeOpener(&file{})
+
+// Open an entry for reading or writing. Several patterns exist for how to interact with entries.
+// - An entry that only supports Read can only be opened ReadOnly.
+// - An entry that supports both Read and Write can be opened in any mode.
+// - An entry that only supports Write can only be opened WriteOnly.
+//
+// When writing and flushing a file, we may call Read on the entry (if it supports Read) even if
+// opened WriteOnly. That only happens when performing a partial write of a *file-like* entry.
 func (f *file) Open(ctx context.Context, req *fuse.OpenRequest, resp *fuse.OpenResponse) (fs.Handle, error) {
-	activity.Record(ctx, "FUSE: Open %v", f)
-
-	// Check for an updated entry in case it has static state.
-	updatedEntry, err := f.refind(ctx)
-	if err != nil {
-		activity.Warnf(ctx, "FUSE: Open errored %v, %v", f, err)
-		return nil, err
-	}
-
-	var fh fileHandle
-	fh.id = f.String()
-
-	// Initiate content request and return a channel providing the results.
-	if plugin.ReadAction().IsSupportedOn(updatedEntry) {
-		activity.Record(ctx, "FUSE: Opened %v", f)
-		if attr := plugin.Attributes(updatedEntry); !attr.HasSize() {
-			// The entry's content size is unknown so open the file in direct
-			// IO mode. This enables FUSE to still read the entry's content so
-			// that built-in tools like cat and grep still work.
-			resp.Flags |= fuse.OpenDirectIO
+	f.mux.Lock()
+	defer f.mux.Unlock()
+	activity.Record(ctx, "FUSE: Open %v: %+v", f, *req)
+
+	if !f.useLocalState() {
+		// Check for an updated entry in case it has static state.
```

Comment:
Why do we need to refresh the entry here?

Reply:
Preloaded external plugin content. I'll mention that in the comment.